kyco/controller.py

```diff
@@ -501,106 +502,71 @@ def add_new_switch(self, switch):
+            napp = module.load_module().Main(controller=self)
+            self.napps[(author, napp_name)] = napp
 
-        Downloads the NApps from the NApp network and install it.
-        TODO: Download or git-clone?
+            for event, listeners in napp._listeners.items():
+                self.events_listeners.setdefault(event, []).extend(listeners)
 
-        Parameters:
-            napp_name (str): Name of the NApp to be installed.
-        """
-        pass
+            napp.start()
 
     def load_napps(self):
-        """Load all NApps installed on the NApps dir."""
-        ignored_path = ['.installed', '__pycache__', '__init__.py']
-        napps_dir = self.options.napps
-
-        try:
-            for author in os.listdir(napps_dir):
-                # Avoid looking at .installed directory
-                if author not in ignored_path:
-                    author_dir = os.path.join(napps_dir, author)
-                    for napp_name in os.listdir(author_dir):
-                        if napp_name not in ignored_path:
-                            full_name = "{}/{}".format(author, napp_name)
-                            log.info("Loading app %s", full_name)
-                            self.load_napp(full_name)
-        except FileNotFoundError as e:
-            log.error("Could not load napps: %s", e)
-
-    def unload_napp(self, napp_name):
-        """Unload a specific NApp based on its name.
-
-        Parameters:
+        """Load all NApps enabled on the NApps dir."""
+        napps = NAppsManager(self.options)
+        for author, napp_name in napps.get_enabled():
+            try:
+                log.info("Loading NApp %s/%s", author, napp_name)
+                self.load_napp(author, napp_name)
+            except FileNotFoundError as e:
+                log.error("Could not load NApp %s/%s: %s", author,
+                          napp_name, e)
+
+    def unload_napp(self, author, napp_name):
+        """Unload a specific NApp.
+
+        Args:
+            author (str): NApp author name.
             napp_name (str): Name of the NApp to be unloaded.
         """
-        napp = self.napps.pop(napp_name)
-        napp.shutdown()
-        # Removing listeners from that napp
-        for event_type, listeners in napp._listeners.items():
-            for listener in listeners:
-                self.events_listeners[event_type].remove(listener)
-            if len(self.events_listeners[event_type]) == 0:
-                self.events_listeners.pop(event_type)
+        napp = self.napps.pop((author, napp_name), None)
+        if napp is None:
+            log.warn('NApp %s/%s was not loaded', author, napp_name)
+        else:
+            napp.shutdown()
+            # Removing listeners from that napp
+            for event_type, napp_listeners in napp._listeners.items():
+                event_listeners = self.events_listeners[event_type]
+                for listener in listeners:
```

Comment:
@cemsbr I think that there is something wrong with the '*listeners*' variables on this block...